src/vs/workbench/common/editor/editorLabels.ts

```diff
@@ -0,0 +1,188 @@
+import { InstantiationType, registerSingleton } from 'vs/platform/instantiation/common/extensions';
+import { createDecorator } from 'vs/platform/instantiation/common/instantiation';
+import { IWorkspaceContextService, IWorkspaceFolder } from 'vs/platform/workspace/common/workspace';
+
+interface IFilenameAttributes {
+	// grand-parent/parent/filename.ext1.ext2 -> parent
+	dirname: string;
+
+	// grand-parent/parent/filename.ext1.ext2 -> [grand-parent, parent]
+	pathFragments: string[];
+
+	// grand-parent/parent/filename.ext1.ext2 -> filename.ext1
+	filename: string;
+
+	// grand-parent/parent/filename.ext1.ext2 -> ext2
+	extname: string;
+}
+
+interface IEditorCustomLabelObject {
+	[key: string]: string;
+}
+
+interface ICustomEditorLabelPattern {
+	pattern: string;
+	template: string;
+}
+
+export class CustomEditorLabelService extends Disposable implements ICustomEditorLabelService {
+
+	readonly _serviceBrand: undefined;
+
+	static readonly SETTING_ID_PATTERNS = 'workbench.editor.label.patterns';
+	static readonly SETTING_ID_ENABLED = 'workbench.editor.label.enabled';
+
+	private readonly _onDidChange = this._register(new Emitter<void>());
+	readonly onDidChange = this._onDidChange.event;
+
+	private patterns: ICustomEditorLabelPattern[] = [];
+	private enabled = true;
+
+	constructor(
+		@IConfigurationService private readonly configurationService: IConfigurationService,
+		@IWorkspaceContextService private readonly workspaceContextService: IWorkspaceContextService,
+	) {
+		super();
+
+		this.storeEnablementState();
+		this.storeCustomPatterns();
+
+		this.registerListernes();
+	}
+
+	private registerListernes(): void {
+		this._register(this.configurationService.onDidChangeConfiguration(e => {
+			// Cache the enabled state
+			if (e.affectsConfiguration(CustomEditorLabelService.SETTING_ID_ENABLED)) {
+				const oldEnablement = this.enabled;
+				this.storeEnablementState();
+				if (oldEnablement !== this.enabled && this.patterns.length > 0) {
+					this._onDidChange.fire();
+				}
+			}
+
+			// Cache the patterns
+			else if (e.affectsConfiguration(CustomEditorLabelService.SETTING_ID_PATTERNS)) {
+				this.storeCustomPatterns();
+				this._onDidChange.fire();
+			}
+		}));
+	}
+
+	private storeEnablementState(): void {
+		this.enabled = this.configurationService.getValue<boolean>(CustomEditorLabelService.SETTING_ID_ENABLED);
+	}
+
+	private storeCustomPatterns(): void {
+		this.patterns = [];
+		const customLabelPatterns = this.configurationService.getValue<IEditorCustomLabelObject>(CustomEditorLabelService.SETTING_ID_PATTERNS);
+		for (const pattern in customLabelPatterns) {
+			const template = customLabelPatterns[pattern];
+			this.patterns.push({ pattern, template });
+		}
+
+		this.patterns.sort((a, b) => this.patternWeight(b.pattern) - this.patternWeight(a.pattern));
+	}
+
+	private patternWeight(pattern: string): number {
+		pattern = pattern.replace('\\\\', '/');
```

Comment:
Fyi glob patterns do not allow backslash, so I think this can be skipped: https://code.visualstudio.com/docs/editor/glob-patterns